src/main/java/edu/harvard/iq/dataverse/api/imports/ImportServiceBean.java

```diff
@@ -268,116 +270,132 @@ public Dataset doImportHarvestedDataset(DataverseRequest dataverseRequest, Harve
             parser.setLenient(true);
-            Dataset ds = parser.parseDataset(obj);
 
-            // For ImportType.NEW, if the metadata contains a global identifier, and it's not a protocol
-            // we support, it should be rejected.
-            // (TODO: ! - add some way of keeping track of supported protocols!)
-            //if (ds.getGlobalId() != null && !ds.getProtocol().equals(settingsService.getValueForKey(SettingsServiceBean.Key.Protocol, ""))) {
-            //    throw new ImportException("Could not register id " + ds.getGlobalId() + ", protocol not supported");
-            //}
-            ds.setOwner(owner);
-            ds.getLatestVersion().setDatasetFields(ds.getLatestVersion().initDatasetFields());
+            if (existingDataset == null) {
+                // Creating a new dataset from scratch:
+                
+                harvestedDataset = parser.parseDataset(obj);
+                harvestedDataset.setOwner(owner);
+                
+                harvestedDataset.setHarvestedFrom(harvestingClient);
+                harvestedDataset.setHarvestIdentifier(harvestIdentifier);
+                
+                harvestedVersion = harvestedDataset.getVersions().get(0);
+            } else {
+                // We already have a dataset with this id in the database.
+                
+                // If this dataset already exists IN ANOTHER COLLECTION
+                // we are just going to skip it!
+                if (existingDataset.getOwner() != null && !owner.getId().equals(existingDataset.getOwner().getId())) {
+                    throw new ImportException("The dataset with the global id " + globalIdString + " already exists, in the dataverse " + existingDataset.getOwner().getAlias() + ", skipping.");
+                }
+                // And if we already have a dataset with this same global id at 
+                // this Dataverse instance, but it is a LOCAL dataset (can happen!), 
+                // we're going to skip it also: 
+                if (!existingDataset.isHarvested()) {
+                    throw new ImportException("A LOCAL dataset with the global id " + globalIdString + " already exists in this dataverse; skipping.");
+                }
+                // For harvested datasets, there should always only be one version.
+                // We will replace the current version with the imported version.
+                // @todo or should we just destroy any extra versions quietly?
+                if (existingDataset.getVersions().size() != 1) {
+                    throw new ImportException("Error importing Harvested Dataset, existing dataset has " + existingDataset.getVersions().size() + " versions");
+                }
+                                
+                harvestedVersion = parser.parseDatasetVersion(obj.getJsonObject("datasetVersion"));
```

Comment:
The next few lines felt kludgy to write... and they are, I can see now how some of the extra logic is not needed and how to streamline things here.